roadmap/roadmap.adoc

```diff
@@ -0,0 +1,158 @@
+Similar to validation, the baseic structure of scoping is very similar in most systems; this _base language_ would condense the common concepts.
+
+===== Typesystem
+We need to calculate the type of a node in several contexts, including validation (e.g. "do the types of these assigned nodes match?") and scoping (e.g. "which members does the thing on the left side of the dot have?").
+
+===== Commenting / Review notes
+For any kind of collaboration, it's very helpful for reviewers to attach their remarks directly to their context -- think of the blue speech bubbles in the right margin of a Word document.
+
+
+==== Access control
+Different users might have different level of access to the whole repository, or parts of it.
+Some users might have read-only access (e.g. via public web), whereas others can write to the repository.{fn-org27}
+
+==== Versioning / Live collaboration
+A repository might support version control and/or live collaboration of users (both features are technically related).
+The repository might have a version control backend like git, or support live collaboration like Google Docs.{fn-org26}
+
+==== Locking
+A repository might support locking the whole repository, partitions, or subtrees.
+The lock might be established with optimistic or pessimistic locking strategies.{fn-org30}
+
+==== Federation
+A repository _A_ might access a different repository _B_, and provide _B_'s contents to _A_'s users.
+The two repositories might be synchronized on a regular schedule, or continuously.{fn-org23}
+
+==== Global symbol index
+A global index of (fully qualified) names can be very useful for any kind of name lookup, for example in parser-based name resolution, scoping, or auto-completion.{fn-org144}
+
+==== Identifier mapping API
+Different clients might refer to the same node by different ids.
+The repository might support them by automatically mapping between the repository's internal id, and the id known to each client.{fn-org94}
+
+==== Language migration API
+Support a common way how to handle language migrations, e.g. by specific messages like "I'm about to migrate this partition" or "You must not use language X in a version older than abc".{fn-org72}
+
+==== Store invalid text typed at arbitrary place
+Given a valid input with invalid text at few places, both tolerant parsers and projectional editors can provide a "mostly correct" AST representation.
+Internally, they have some way of keeping the invalid text around.
+We could support a standard way to store such invalid text in the repository.{fn-org62}
+
+==== Define a standard way to represent model diffs
+Calculating the difference between two sets of nodes is common, complex, and potentially expensive task.
+It might be useful to have a standard format to exchange the difference, and the <<delta-protocol>> could provide a solid base.{fn-org40}
```

Comment:
We might want to look at [EMF Compare](https://eclipse.dev/emf/compare/) for this. We could reimplement this, or re-use EMF Compare by exporting to EMF(/Ecore), running the comparison, and importing the result from EMF(/Ecore) to LionWeb again.

Reply:
True. EMF Compare might have some interesting insights that could inform the precise protocol as well, though.